### PR TITLE
Erc20 and WrappedCSPR modules

### PR DESCRIPTION
**Pull request by kpob**

### CODE REVIEW

File: modules/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "odra-modules"
+version = "0.1.0"
+edition = "2021"
+authors = ["Jakub Płaskonka <kuba@odra.dev>", "Krzysztof Pobiarżyn <krzysztof@odra.dev>", "Maciej Zieliński <maciej@odra.dev>"]
+license = "MIT"
+repository = "https://github.com/odradev/odra"
+description = "Collection of reusable Odra modules"
```
Comment (zie1ony) [inline modules/Cargo.toml]:
```suggestion
description = "Collection of reusable Odra modules."
```

File: modules/src/erc20.rs [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    errors::Error,
+    events::{Approval, Transfer}
+};
+
+#[odra::module]
+pub struct Erc20 {
+    decimals: Variable<u8>,
+    symbol: Variable<String>,
+    name: Variable<String>,
+    total_supply: Variable<Balance>,
```
Comment (zie1ony) [inline modules/src/erc20.rs]:
```suggestion
    total_supply: Variable<U256>,
```
The type for balance must to be `U256`. Reasons:
- we need to be as close to original standard as possible,
- this will be used by ERC20-based assets biridged into Casper.

File: modules/src/erc20.rs [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        Approval {
+            owner,
+            spender,
+            value: amount
+        }
+        .emit();
+    }
+
+    pub fn name(&self) -> String {
+        self.name.get_or_default()
```
Comment (zie1ony) [inline modules/src/erc20.rs]:
```suggestion
        self.name.unwrap_or_revert_with(Error::NameNotSet)
```
would be more strict in case of module not being initialized.

File: modules/src/erc20.rs [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+#[odra::module]
+impl Erc20 {
+    #[odra(init)]
+    pub fn init_with_supply(
+        &mut self,
+        symbol: String,
+        name: String,
+        decimals: u8,
+        initial_supply: Balance
```
Comment (zie1ony) [inline modules/src/erc20.rs]: Wrap in `Option`.

File: modules/src/erc20.rs [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+        self.allowances.subtract(&key, amount);
+        Approval {
+            owner,
+            spender,
+            value: amount
+        }
+        .emit();
+    }
+
+    pub fn increase_total_supply(&mut self, amount: Balance) {
```
Comment (zie1ony) [inline modules/src/erc20.rs]: Remove those 4 methods.

File: modules/src/erc20.rs [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        }
+    }
+}
+
+#[cfg(all(test, feature = "mock-vm"))]
+mod tests {
+
+    use odra::{assert_events, test_env, types::Balance};
+
+    use crate::erc20::events::Transfer;
```
Comment (zie1ony) [inline modules/src/erc20.rs]: Merge imports.

File: modules/src/wrapped_native.rs [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+#[cfg(all(test, feature = "mock-vm"))]
+mod tests {
+
+    use odra::{
+        assert_events, test_env,
+        types::{Address, Balance, OdraError, VmError}
+    };
+
+    use crate::{
```
Comment (zie1ony) [inline modules/src/wrapped_native.rs:119]: merge imports

File: modules/src/erc20.rs [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+            let mut erc20 = setup();
+
+            // When the transfer amount exceeds the sender balance.
+            let recipient = test_env::get_account(1);
+            let amount = Balance::from(INITIAL_SUPPLY) + Balance::one();
+
+            // Then an error occurs.
+            erc20.transfer(recipient, amount)
+        });
+    }
```
Comment (zie1ony) [inline modules/src/erc20.rs:288]: Add transfer_from + approve tests.

Comment (zie1ony) [reply modules/src/erc20.rs:288]: https://github.com/odradev/odra-examples/blob/0.2.0/src/erc20.rs#L212

File: modules/src/wrapped_native.rs [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+    pub fn init(&mut self) {
+        let metadata = contract_env::native_token_metadata();
+        self.erc20
+            .init(metadata.symbol, metadata.name, metadata.decimals);
+    }
+
+    #[odra(payable)]
+    pub fn deposit(&mut self) {
+        let caller = contract_env::caller();
+        let amount = contract_env::attached_value();
```
Comment (zie1ony) [inline modules/src/wrapped_native.rs]: Parse U512 to U256 and revert if too large.

File: modules/src/wrapped_native.rs [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+
+        self.erc20.mint(caller, amount);
```
Comment (zie1ony) [inline modules/src/wrapped_native.rs:32]: Add issue to test U256 overflow.

File: modules/src/wrapped_native.rs [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+            token.total_supply(),
+            (deposit_amount + deposit_amount).into()
+        );
+        // Then events were emitted.
+        assert_events!(token, Transfer, Deposit, Transfer, Deposit);
+    }
+
+    #[test]
+    fn test_deposit_amount_exceeding_account_balance() {
+        //TODO: Consider what really should happen here.
```
Comment (zie1ony) [inline modules/src/wrapped_native.rs]:
```suggestion

```

File: odra-casper/shared/src/native_token.rs [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+        Self::new()
+    }
+}
+
+impl NativeTokenMetadata {
+    pub fn new() -> Self {
+        Self {
+            name: String::from("Casper"),
+            symbol: String::from("CSPR"),
+            decimals: 8
```
Comment (zie1ony) [inline odra-casper/shared/src/native_token.rs]:
```suggestion
            decimals: 9
```

File: modules/src/wrapped_native.rs [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+pub struct WrappedNativeToken {
+    erc20: Erc20
+}
+
+#[odra::module]
+impl WrappedNativeToken {
+    #[odra(init)]
+    pub fn init(&mut self) {
+        let metadata = contract_env::native_token_metadata();
+        self.erc20
```
Comment (zie1ony) [inline modules/src/wrapped_native.rs]: add `format!("Wrapped {}", metadata.name)` and for symbol `W{}`

File: mock-vm/backend/src/native_token.rs [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+    pub symbol: String,
+    pub decimals: u8
+}
+
+impl NativeTokenMetadata {
+    pub(crate) fn new() -> Self {
+        Self {
+            name: String::from("Plascoin"),
+            symbol: String::from("PLS"),
+            decimals: 8
```
Comment (zie1ony) [inline mock-vm/backend/src/native_token.rs]:
```suggestion
            decimals: 2
```

**Review by zie1ony — CHANGES_REQUESTED** [review]

**Review by zie1ony — APPROVED** [review]
